log-replication/src/main/java/org/corfudb/logreplication/fsm/LogReplicationFSM.java

```diff
@@ -0,0 +1,352 @@
+        try {
+            if (state.getType() == LogReplicationStateType.STOPPED) {
+                log.info("Log Replication State Machine has been stopped. No more events will be processed.");
+                return;
+            }
+
+            // TODO (Anny): consider strategy for continuously failing snapshot sync (never ending cancellation)
+            //   Block until an event shows up in the queue.
+            LogReplicationEvent event = eventQueue.take();
+            if (event.getType() != LogReplicationEventType.LOG_ENTRY_SYNC_CONTINUE) {
+                log.info("Log Replication FSM consume event {}", event);
+            }
+
+            if (event.getType() == LogReplicationEventType.LOG_ENTRY_SYNC_REPLICATED) {
+                if (state.getType() == LogReplicationStateType.IN_LOG_ENTRY_SYNC &&
+                        state.getTransitionEventId().equals(event.getMetadata().getRequestId())) {
+                    log.debug("Log Entry Sync ACK, update last ack timestamp to {}", event.getMetadata().getSyncTimestamp());
+                    ackedTimestamp = event.getMetadata().getSyncTimestamp();
+                }
+            } else {
+                if (event.getType() == LogReplicationEventType.SNAPSHOT_SYNC_COMPLETE) {
+                    // Verify it's for the same request, as that request could've been canceled and was received later
+                    if (state.getType() == LogReplicationStateType.IN_SNAPSHOT_SYNC &&
+                            state.getTransitionEventId().equals(event.getMetadata().getRequestId())) {
+                        log.debug("Snapshot Sync ACK, update last ack timestamp to {}", event.getMetadata().getSyncTimestamp());
+                        baseSnapshot = event.getMetadata().getSyncTimestamp();
+                        ackedTimestamp = baseSnapshot;
+                    }
+                }
+
+                try {
+                    LogReplicationState newState = state.processEvent(event);
+                    log.trace("Transition from {} to {}", state, newState);
+                    transition(state, newState);
+                    state = newState;
+                    numTransitions.setValue(numTransitions.getValue() + 1);
+                } catch (IllegalTransitionException illegalState) {
+                    log.debug("Illegal log replication event {} when in state {}", event.getType(), state.getType());
+                }
+            }
+
+            //For testing purpose to notify the event generator the stop of the event.
+            if (event.getType() == LogReplicationEventType.REPLICATION_STOP) {
+                synchronized (event) {
+                    event.notifyAll();
+                }
+            }
+
+            // Consume one event in the queue and re-submit, this is done so events are consumed in
+            // a round-robin fashion for the case of multi-site replication.
+            logReplicationFSMConsumer.submit(this::consume);
```

Comment:
instead of this, you can configure logReplicationFSMConsumer to run at a customizable interval using its constructor parameters.